navapbc/labs-gen-ai-experiments
# feat: Add summarizing and quoting capabilities to 05-assistive-chatbot

## Ticket

Partly resolves https://navalabs.atlassian.net/browse/DST-218

## Changes

Populate prototype code in `05-assistive-chatbot` with the summarizing and quoting capabilities from `02-household-queries` chatbot.

Update handling `settings` for LLM clients and did some general refactoring.

## Context for reviewers

This provides baseline summarizing and quoting capabilities for the prototype.

## Testing

Testing instructions and expected behavior:
1. `pip install -r requirements.txt`
1. Update `.env`
2. To clear and populate the DB, run: `rm -rf chroma_db && ./ingest-guru-cards.py`
1. Try it on the command line, for example:
```
CHATBOT_LOG_LEVEL=INFO CHAT_ENGINE=Summaries    LLM_MODEL_NAME='openai :: gpt-3.5-turbo-instruct' LLM_MODEL_NAME_2='openai :: gpt-3.5-turbo-instruct' RETRIEVE_K=2 ./cmdline.py

CHATBOT_LOG_LEVEL=INFO CHAT_ENGINE=Summaries-DSPy LLM_MODEL_NAME='dspy :: gpt-3.5-turbo-instruct' LLM_MODEL_NAME_2='openai :: gpt-3.5-turbo-instruct' RETRIEVE_K=2 ./cmdline.py
```
1. Try it via the web app (To ensure summaries are generated correctly, pick `*instruct` model for both LLMs):
```
./chatbot-chainlit.py
```
   - Open a browser to http://localhost:8000/
   - Open the `Settings panel` by clicking on the icon to the left of the bottom text input field.
   - Choose "Summaries" as the `Chat Mode` (synonymous with `CHAT_ENGINE` for the `cmdline.py` app) and pick `openai :: gpt-3.5-turbo-instruct` as the `Primary LLM` and `LLM Model for summarizer`. Adjust the two `Temperature...` settings to play with the responses.
   - click `Confirm`
   - type in any question and see derived questions, Guru cards, their summaries, and quotes:
<img width="726" alt="image" src="https://github.com/navapbc/labs-gen-ai-experiments/assets/55255674/8442b515-3598-4a01-8887-e570cf01520b">
